// vite-plugin-strip@1.1.0/dist/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);
var regexCharsRegex = /[-[\]{}()*+?.,\\^$\\/|#]/g;
var escapeRegexChars = (comment) => {
  return comment && regexCharsRegex.test(comment) ? comment.replace(regexCharsRegex, "\\$&") : comment;
};
var DEFAULT_OPTIONS = {
  enabled: true,
  domainList: [],
  start: "devBlock:start",
  end: "devBlock:end"
};
var stripPlugin = (options) => {
  const normalizedOptions = {
    ...DEFAULT_OPTIONS,
    ...options
  };
  return {
    name: "vite-plugin-strip",
    enforce: "pre",
    transform(code, id) {
      if (!normalizedOptions.enabled || !/\.([jt]sx?)$/.test(id)) {
        return code;
      }
      if (/\/src\/main\.(js|ts)/.test(id)) {
        const domainList = normalizedOptions.domainList;
        if (Array.isArray(domainList) && domainList.length > 0) {
          const disableStr = JSON.stringify(domainList);
          code = `${code}
 if (${disableStr}.indexOf(location.host) !== -1) { console.log = function () {} }`;
        }
      }
      const startComment = escapeRegexChars(normalizedOptions.start);
      const endComment = escapeRegexChars(normalizedOptions.end);
      const startWhitespaceMatcher = "";
      const endWhitespaceMatcher = "";
      const regexPattern = new RegExp(
        startWhitespaceMatcher + "\\/\\* ?" + startComment + " ?\\*\\/[\\s\\S]*?\\/\\* ?" + endComment + " ?\\*\\/" + endWhitespaceMatcher,
        "g"
      );
      return code.replace(regexPattern, "");
    }
  };
};
var src_default = stripPlugin;
