// es-toolkit@1.36.0/dist/predicate/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const isPromise = require('../_chunk/isPromise-CxqI1v.js');

function isBoolean(x) {
    return typeof x === 'boolean';
}

function isError(value) {
    return value instanceof Error;
}

function isString(value) {
    return typeof value === 'string';
}

exports.isArrayBuffer = isPromise.isArrayBuffer;
exports.isBlob = isPromise.isBlob;
exports.isBrowser = isPromise.isBrowser;
exports.isBuffer = isPromise.isBuffer;
exports.isDate = isPromise.isDate;
exports.isEqual = isPromise.isEqual;
exports.isEqualWith = isPromise.isEqualWith;
exports.isFile = isPromise.isFile;
exports.isFunction = isPromise.isFunction;
exports.isJSON = isPromise.isJSON;
exports.isJSONArray = isPromise.isJSONArray;
exports.isJSONObject = isPromise.isJSONObject;
exports.isJSONValue = isPromise.isJSONValue;
exports.isLength = isPromise.isLength;
exports.isMap = isPromise.isMap;
exports.isNil = isPromise.isNil;
exports.isNode = isPromise.isNode;
exports.isNotNil = isPromise.isNotNil;
exports.isNull = isPromise.isNull;
exports.isPlainObject = isPromise.isPlainObject;
exports.isPrimitive = isPromise.isPrimitive;
exports.isPromise = isPromise.isPromise;
exports.isRegExp = isPromise.isRegExp;
exports.isSet = isPromise.isSet;
exports.isSymbol = isPromise.isSymbol;
exports.isTypedArray = isPromise.isTypedArray;
exports.isUndefined = isPromise.isUndefined;
exports.isWeakMap = isPromise.isWeakMap;
exports.isWeakSet = isPromise.isWeakSet;
exports.isBoolean = isBoolean;
exports.isError = isError;
exports.isString = isString;
