// es-toolkit@1.36.0/dist/string/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const reverseString = require('../_chunk/reverseString-BixeGz.js');

function startCase(str) {
    const words = reverseString.words(str.trim());
    let result = '';
    for (let i = 0; i < words.length; i++) {
        const word = words[i];
        if (result) {
            result += ' ';
        }
        result += word[0].toUpperCase() + word.slice(1).toLowerCase();
    }
    return result;
}

exports.camelCase = reverseString.camelCase;
exports.capitalize = reverseString.capitalize;
exports.constantCase = reverseString.constantCase;
exports.deburr = reverseString.deburr;
exports.escape = reverseString.escape;
exports.escapeRegExp = reverseString.escapeRegExp;
exports.kebabCase = reverseString.kebabCase;
exports.lowerCase = reverseString.lowerCase;
exports.lowerFirst = reverseString.lowerFirst;
exports.pad = reverseString.pad;
exports.pascalCase = reverseString.pascalCase;
exports.reverseString = reverseString.reverseString;
exports.snakeCase = reverseString.snakeCase;
exports.trim = reverseString.trim;
exports.trimEnd = reverseString.trimEnd;
exports.trimStart = reverseString.trimStart;
exports.unescape = reverseString.unescape;
exports.upperCase = reverseString.upperCase;
exports.upperFirst = reverseString.upperFirst;
exports.words = reverseString.words;
exports.startCase = startCase;
