// tdesign-vue-next@1.11.5/es/_chunks/dep-251d5b63.mjs
/**
 * tdesign v1.11.5
 * (c) 2025 tdesign
 * @license MIT
 */

var EKeyboardDirection = /* @__PURE__ */function (EKeyboardDirection2) {
  EKeyboardDirection2[EKeyboardDirection2["left"] = 37] = "left";
  EKeyboardDirection2[EKeyboardDirection2["up"] = 38] = "up";
  EKeyboardDirection2[EKeyboardDirection2["right"] = 39] = "right";
  EKeyboardDirection2[EKeyboardDirection2["down"] = 40] = "down";
  return EKeyboardDirection2;
}(EKeyboardDirection || {});
var ARROW_DOWN_REG = /^ArrowDown$/i;
var ARROW_UP_REG = /^ArrowUp$/i;
var ARROW_LEFT_REG = /^ArrowLeft$/i;
var ARROW_RIGHT_REG = /^ArrowRight$/i;
var ESCAPE_REG = /^Escape$/i;
var SPACE_REG = /^Space$/i;
var ENTER_REG = /^Enter$/i;
var SHIFT_REG = /^(Shift|ShiftLeft|ShiftRight)$/i;
var CLEAR_REG = /^KeyC$/i;
var ALL_REG = /^(KeyA|KeyL)$/i;
var CHECKED_CODE_REG = /^(Enter|Space)$/i;

export { ARROW_UP_REG as A, CHECKED_CODE_REG as C, ENTER_REG as E, SPACE_REG as S, ARROW_DOWN_REG as a, SHIFT_REG as b, ESCAPE_REG as c, CLEAR_REG as d, ALL_REG as e, ARROW_LEFT_REG as f, ARROW_RIGHT_REG as g };
